Add JsonValidation for event messages

jasperpeeters opened the pull request:
Adding the last missing schema validation for the Event messages received via the websocket subscriptions.

As the schema for the event messages is a slightly different case (not reachable via the `QueryAPI.raml`) in comparison to the other schemas, I have to load the schema and validate manually at the moment.

Currently only SYNC messages are validated, but as they have both (`pre` and `post`) messages filled, I think this might be enough and there is no need to validate the messages for MODIFIED, ADD and DELETED as well. Thoughts?

andrewbonney commented:
Given what's already tested I suspect that will catch most things. Is there any harm in adding them for each of the other cases beyond it perhaps taking a little longer?

garethsb-sony commented:
I'm trying to run this on Windows and am getting 

```
jsonschema.exceptions.RefResolutionError: <urlopen error [WinError 2] The system cannot find the file specified: '\\node.json'>
```

This is obviously related to #5, but I can't see quite how to apply that solution here?
(And also to the similar code in IS0501Test.check_staged_complies_with_constraints presumably.)

andrewbonney commented:
I'm very tempted to replace these RefResolver instances with jsonref as used by ramlfications (assuming it actually works better). Each time this comes up it seems to require a subtly different solution for no obvious reason.

jasperpeeters commented:
Ok, I got a fix now, which seems to work on Windows and Linux. It's using a slightly modified approach as the patch function from #5.
I added a new function to `GenericTest.py`:
```python
    def load_resolved_schema(self, api_name, file_name):
        """
        Parses JSON as well as resolves any `$ref`s, including references to
        local files and remote (HTTP/S) files.
        """
        base_path = os.path.abspath(os.path.join(self.apis[api_name]["spec_path"], "APIs/schemas/"))
        json_file = os.path.join(base_path, file_name)

        if not base_path.endswith("/"):
            base_path = base_path + "/"
        if os.name == "nt":
            base_path = "file:///" + base_path.replace('\\', '/')
        else:
            base_path = "file://" + base_path

        loader = jsonref.JsonLoader(cache_results=False)

        with open(json_file, "r") as f:
            schema = jsonref.load(f, base_uri=base_path, loader=loader, jsonschema=True)

        return schema
```

and can call this in the affected `test_31` with: 
```python
            # Load schema
            if self.is04_reg_utils.compare_api_version(api["version"], "v1.0") == 0:
                schema = self.load_resolved_schema("query", "queryapi-v1.0-subscriptions-websocket.json")
            else:
                schema = self.load_resolved_schema("query", "queryapi-subscriptions-websocket.json")
```

which allows the validation later on via: 
```python
                    try:
                        Draft4Validator(schema).validate(json.loads(message))
                    except ValidationError as e:
                        return test.FAIL("Received event message is invalid: {}".format(str(e)))
```

However, I'm not sure how to approach this now, as the problem seems to occur on other places as well, as @garethsb-sony mentioned.

If you think this solution works and is in general reusable, I can push my changes.

andrewbonney commented:
Assuming I've not missed anything I'd be keen to see this jsonref version used as the primary way to resolve any schema references. It would be great to use it wherever a 'RefResolver' is currently used, and as a replacement for '_deref_schema' in Specification.py. It might need to go into TestHelper.py as a general utility function to make it available to all of the classes, but hopefully it would be a relatively straightforward drop-in replacement for the existing mechanisms.

garethsb-sony commented:
Am I right in thinking that doing it this way effectively builds the complete schema in advance of actually using it; so no resolution of ``$ref`` happens during ``validate`` at all?

andrewbonney commented:
That's my understanding

jasperpeeters commented:
I now moved the `load_resolved_schema` func to the `TestHelper` class. Additionally I added the validation for the remaining event types (modified, removed and added).

garethsb-sony commented:
Only thing, @jasperpeeters, @andrewbonney, did we ought to refactor ``_deref_schema`` in Specification.py at the same time?

andrewbonney commented:
Yes, if we can get rid of other schema dereferencing stuff that'd be great. I haven't tested it yet, but it looks good subject to a couple of minor suggestions above.
